src/Zone.js

```diff
@@ -15,41 +15,62 @@ import UXBase from './UXBase';
  * }
  */
 export default class Zone extends UXBase {
-	// Name used for UserTiming measures
-	measureName = this.props.name;
-
-	// Create a new `ExpectedMark` for each mark
 	marks = this.props.marks.map(markName => {
 		const mark = ExpectedMark.create(markName);
 
-		mark.onComplete(completeMark => {
+		const listener = completeMark => {
 			// pass the event upstream
 			this.props.onMark(markName);
-			if (this.marks.every(m => m.marked)) {
+			if (this.marks.every(({ marked }) => mark.marked)) {
 				this.measure(markName);
 			}
-		});
+		};
+		mark.addOnMarkListener(listener);
 
-		return mark;
+		return { mark, listener };
```

Comment:
instead returning just the mark, need to keep a reference to the `listener` so that it can be removed when the Zone is destroyed. Without the ability to remove the specific listener added by this Zone, there is a memory leak and potential thrown exception because listener can outlive its parent zone

Reply:
follow-on: I've got some ideas about making this much simpler, but it will require some experimentation. Essentially, all these `listener` callbacks are almost identical, so we should be able to just pass in a single listener that gets used whenever any mark is recorded, but the implications of that architecture need some evaluation after these core changes are settled